File: solution/backtracking/15655/main.py
# Authored by : tony9402
# Co-authored by : -
# Link : http://boj.kr/d4f609a8ba2e40e8b578c37f3d21d344
import sys

def input():
    return sys.stdin.readline().rstrip()

N, M = map(int, input().split())
arr  = sorted(list(map(int, input().split())))
choose = [ 0 for _ in range(10) ]
used   = [ 0 for _ in range(10) ]

def dfs(idx, cnt):
    global N, M
    if cnt == M:
        for idx in range(cnt):
            print(arr[choose[idx]], end=' ')
        print()
        return

    for i in range(idx,N):
        if used[i]:
            continue
        used[i] = 1
        choose[cnt] = i
        dfs(i + 1, cnt + 1)
        used[i] = 0

dfs(0, 0)

File: solution/backtracking/15655/main.cpp
// Authored by : tony9402
// Co-authored by : -
// Link : http://boj.kr/cb5e9a75972a4ac59cc7b4428d07f2fa
#include<bits/stdc++.h>

using namespace std;

int arr[10], choose[10], N, M;
bool used[10];

void dfs(int idx, int cnt) {
    if(cnt == M){
        for(int i=0;i<M;i++) {
            cout << arr[choose[i]] << ' ';
        }
        cout << '\n';
        return;
    }

    for(int i=idx;i<N;i++) {
        if(used[i]) continue;
        used[i] = true;
        choose[cnt] = i;
        dfs(i + 1, cnt + 1);
        used[i] = false;
    }
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0);

    cin >> N >> M;
    for(int i=0;i<N;i++) cin >> arr[i];
    sort(arr, arr + N);
    dfs(0, 0);

    return 0;
}
